### DISCUSSION TITLE
Where to download PhonemeDataset

### DISCUSSION THREAD

**Discussion by DennisMinn** [coqui-ai/TTS#2301] [General Q&A] [answered]
I am following [Training a Model](https://tts.readthedocs.io/en/latest/training_a_model.html), but modified `dataset_config` to match my own dataset. 

When I call `trainer.fit()` I get an error on line 585 `cache_path = os.path.join(self.cache_path, file_name + file_ext)` from `TTS/tts/datasets/dataset.py` because `self.cache_path = None`. 

[LJ Speech Dataset](https://keithito.com/LJ-Speech-Dataset/) doesn't mention including phoneme data, where would I download the phoneme dataset for `self.cache_path`?

**TL;DR, where to download phoneme dataset for the argument `cache_path` under the [PhonemeDataset](https://github.com/coqui-ai/TTS/blob/6e3f74fc2927892b5575d33e4df757e6d9f19fd3/TTS/tts/datasets/dataset.py#L565) class?**

Traceback Error
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/trainer/trainer.py:1591, in Trainer.fit(self)
   1590 try:
-> 1591     self._fit()
   1592     if self.args.rank == 0:

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/trainer/trainer.py:1544, in Trainer._fit(self)
   1543 if not self.skip_train_epoch and not self.start_with_eval:
-> 1544     self.train_epoch()
   1545 if self.config.run_eval:

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/trainer/trainer.py:1308, in Trainer.train_epoch(self)
   1307 batch_num_steps = len(self.train_loader)
-> 1308 for cur_step, batch in enumerate(self.train_loader):
   1309     _, _ = self.train_step(batch, batch_num_steps, cur_step, loader_start_time)

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/torch/utils/data/dataloader.py:628, in _BaseDataLoaderIter.__next__(self)
    627     self._reset()  # type: ignore[call-arg]
--> 628 data = self._next_data()
    629 self._num_yielded += 1

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/torch/utils/data/dataloader.py:1333, in _MultiProcessingDataLoaderIter._next_data(self)
   1332 del self._task_info[idx]
-> 1333 return self._process_data(data)

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/torch/utils/data/dataloader.py:1359, in _MultiProcessingDataLoaderIter._process_data(self, data)
   1358 if isinstance(data, ExceptionWrapper):
-> 1359     data.reraise()
   1360 return data

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/torch/_utils.py:543, in ExceptionWrapper.reraise(self)
    542     raise RuntimeError(msg) from None
--> 543 raise exception

TypeError: Caught TypeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/home/dminn/Projects/moistnet/moist_venv/lib/python3.8/site-packages/torch/utils/data/_utils/worker.py", line 302, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/dminn/Projects/moistnet/moist_venv/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 58, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/dminn/Projects/moistnet/moist_venv/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 58, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/dminn/Projects/moistnet/moist_venv/lib/python3.8/site-packages/TTS/tts/datasets/dataset.py", line 187, in __getitem__
    return self.load_data(idx)
  File "/home/dminn/Projects/moistnet/moist_venv/lib/python3.8/site-packages/TTS/tts/datasets/dataset.py", line 237, in load_data
    token_ids = self.get_token_ids(idx, item["text"])
  File "/home/dminn/Projects/moistnet/moist_venv/lib/python3.8/site-packages/TTS/tts/datasets/dataset.py", line 220, in get_token_ids
    token_ids = self.get_phonemes(idx, text)["token_ids"]
  File "/home/dminn/Projects/moistnet/moist_venv/lib/python3.8/site-packages/TTS/tts/datasets/dataset.py", line 203, in get_phonemes
    out_dict = self.phoneme_dataset[idx]
  File "/home/dminn/Projects/moistnet/moist_venv/lib/python3.8/site-packages/TTS/tts/datasets/dataset.py", line 572, in __getitem__
    ids = self.compute_or_load(string2filename(item["audio_unique_name"]), item["text"], item["language"])
  File "/home/dminn/Projects/moistnet/moist_venv/lib/python3.8/site-packages/TTS/tts/datasets/dataset.py", line 585, in compute_or_load
    cache_path = os.path.join(self.cache_path, file_name + file_ext)
  File "/usr/lib/python3.8/posixpath.py", line 76, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType


During handling of the above exception, another exception occurred:

SystemExit                                Traceback (most recent call last)
    [... skipping hidden 1 frame]

Cell In[180], line 1
----> 1 trainer.fit()

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/trainer/trainer.py:1612, in Trainer.fit(self)
   1611 traceback.print_exc()
-> 1612 sys.exit(1)

SystemExit: 1

During handling of the above exception, another exception occurred:

AssertionError                            Traceback (most recent call last)
    [... skipping hidden 1 frame]

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/IPython/core/interactiveshell.py:2042, in InteractiveShell.showtraceback(self, exc_tuple, filename, tb_offset, exception_only, running_compiled_code)
   2039 if exception_only:
   2040     stb = ['An exception has occurred, use %tb to see '
   2041            'the full traceback.\n']
-> 2042     stb.extend(self.InteractiveTB.get_exception_only(etype,
   2043                                                      value))
   2044 else:
   2045     try:
   2046         # Exception classes can customise their traceback - we
   2047         # use this in IPython.parallel for exceptions occurring
   2048         # in the engines. This should return a list of strings.

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/IPython/core/ultratb.py:585, in ListTB.get_exception_only(self, etype, value)
    577 def get_exception_only(self, etype, value):
    578     """Only print the exception type and message, without a traceback.
    579 
    580     Parameters
   (...)
    583     value : exception value
    584     """
--> 585     return ListTB.structured_traceback(self, etype, value)

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/IPython/core/ultratb.py:452, in ListTB.structured_traceback(self, etype, evalue, etb, tb_offset, context)
    449     chained_exc_ids.add(id(exception[1]))
    450     chained_exceptions_tb_offset = 0
    451     out_list = (
--> 452         self.structured_traceback(
    453             etype, evalue, (etb, chained_exc_ids),
    454             chained_exceptions_tb_offset, context)
    455         + chained_exception_message
    456         + out_list)
    458 return out_list

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/IPython/core/ultratb.py:1118, in AutoFormattedTB.structured_traceback(self, etype, value, tb, tb_offset, number_of_lines_of_context)
   1116 else:
   1117     self.tb = tb
-> 1118 return FormattedTB.structured_traceback(
   1119     self, etype, value, tb, tb_offset, number_of_lines_of_context)

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/IPython/core/ultratb.py:1012, in FormattedTB.structured_traceback(self, etype, value, tb, tb_offset, number_of_lines_of_context)
   1009 mode = self.mode
   1010 if mode in self.verbose_modes:
   1011     # Verbose modes need a full traceback
-> 1012     return VerboseTB.structured_traceback(
   1013         self, etype, value, tb, tb_offset, number_of_lines_of_context
   1014     )
   1015 elif mode == 'Minimal':
   1016     return ListTB.get_exception_only(self, etype, value)

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/IPython/core/ultratb.py:865, in VerboseTB.structured_traceback(self, etype, evalue, etb, tb_offset, number_of_lines_of_context)
    856 def structured_traceback(
    857     self,
    858     etype: type,
   (...)
    862     number_of_lines_of_context: int = 5,
    863 ):
    864     """Return a nice text document describing the traceback."""
--> 865     formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,
    866                                                            tb_offset)
    868     colors = self.Colors  # just a shorthand + quicker name lookup
    869     colorsnormal = colors.Normal  # used a lot

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/IPython/core/ultratb.py:799, in VerboseTB.format_exception_as_a_whole(self, etype, evalue, etb, number_of_lines_of_context, tb_offset)
    796 assert isinstance(tb_offset, int)
    797 head = self.prepare_header(etype, self.long_header)
    798 records = (
--> 799     self.get_records(etb, number_of_lines_of_context, tb_offset) if etb else []
    800 )
    802 frames = []
    803 skipped = 0

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/IPython/core/ultratb.py:854, in VerboseTB.get_records(self, etb, number_of_lines_of_context, tb_offset)
    848     formatter = None
    849 options = stack_data.Options(
    850     before=before,
    851     after=after,
    852     pygments_formatter=formatter,
    853 )
--> 854 return list(stack_data.FrameInfo.stack_data(etb, options=options))[tb_offset:]

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/stack_data/core.py:578, in FrameInfo.stack_data(cls, frame_or_tb, options, collapse_repeated_frames)
    562 @classmethod
    563 def stack_data(
    564         cls,
   (...)
    568         collapse_repeated_frames: bool = True
    569 ) -> Iterator[Union['FrameInfo', RepeatedFrames]]:
    570     """
    571     An iterator of FrameInfo and RepeatedFrames objects representing
    572     a full traceback or stack. Similar consecutive frames are collapsed into RepeatedFrames
   (...)
    576     and optionally an Options object to configure.
    577     """
--> 578     stack = list(iter_stack(frame_or_tb))
    580     # Reverse the stack from a frame so that it's in the same order
    581     # as the order from a traceback, which is the order of a printed
    582     # traceback when read top to bottom (most recent call last)
    583     if is_frame(frame_or_tb):

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/stack_data/utils.py:97, in iter_stack(frame_or_tb)
     95 while frame_or_tb:
     96     yield frame_or_tb
---> 97     if is_frame(frame_or_tb):
     98         frame_or_tb = frame_or_tb.f_back
     99     else:

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/stack_data/utils.py:90, in is_frame(frame_or_tb)
     89 def is_frame(frame_or_tb: Union[FrameType, TracebackType]) -> bool:
---> 90     assert_(isinstance(frame_or_tb, (types.FrameType, types.TracebackType)))
     91     return isinstance(frame_or_tb, (types.FrameType,))

File ~/Projects/moistnet/moist_venv/lib/python3.8/site-packages/stack_data/utils.py:176, in assert_(condition, error)
    174 if isinstance(error, str):
    175     error = AssertionError(error)
--> 176 raise error

AssertionError: 
```

**Comment by Ca-ressemble-a-du-fake** [comment, marked as answer]
Hi,

In my projects I only modify the recipe not the config. In the recipe you've got a line 'phoneme_cache_path=os.path.join(output_path, "phoneme_cache")'. This is where the phoneme cache will be generated from your dataset (automatically).

 TLDR So you should not have to download anything it will be generated automatically.
